File: calendar1_R&A_Algorithm/HPN C# R&A CODE/HpnCalendarTools/HpnCalendarTools/hpn/cs/xml/client/HostUrl.cs
using System;
using hpn.settings;
using System.Text.RegularExpressions;

namespace hpn.cs.xml.client
{
    public class HostUrl
    {
        
        private String hostUrlAddress;
	    private int port;
	
	    public HostUrl(String hostUrl)
	    {
		    this.setHostUrl(hostUrl);
		    this.setPort(DefaultPort.portNumber); //Default port for all hostess is 8080.
	    }
	    public HostUrl(String hostUrl, int port)
	    {
		    this.setHostUrl(hostUrl);
		    this.setPort(port);
	    }
	    public HostUrl(int port, String ipv4Address)
	    {
		    this.setHostUrl("http://"+ipv4Address+"/");
		    this.setPort(port);
	    }
	    internal HostUrl(int port) {
            this.setHostUrl("http://" + MachinIdentification.getIpAddress() + "/"); //For sending requests to the current machine
            this.setPort(port != -1 ? port : MachinIdentification.getPort()); //Default port for all hostess is 8080.
	    }
	    public String getFullUrl()
	    {
        	    Uri uri = new Uri(this.hostUrlAddress);
        	    return uri.Scheme + "://" + uri.Host+ ":" + this.port + "/";
	    }
	    public String getHostUrl() {
		    return this.hostUrlAddress;
	    }

	    public void setHostUrl(String hostUrlAddress) 
	    {
		    Regex regex = new Regex(@"^http://([\w-]+\.)+[\w-]+(/[\w- ./?%&=]*)?");
            if (regex.IsMatch(hostUrlAddress)) //url validation
		    {
			    this.hostUrlAddress = hostUrlAddress;
		    } 
		    else 
		    {
                throw new System.FormatException("The URL address of the host [" + hostUrlAddress + "] is invalid.");
		    }
	    }

	    public int getPort() 
	    {
		    return this.port;
	    }

	    public void setPort(int port) 
	    {
		    if(port>1024 && port<= 65535)
		    {
			    this.port = port;
		    }
		    else 
		    {
		           throw new System.ArgumentOutOfRangeException("The port number of the host ["+port+"] is invalid. It must be between 1025 and  65535.");
		    }
	    }
	    
	    public override String ToString(){
		    return "URL:["+this.getHostUrl()+"] Port:["+this.getPort()+"]\n";
	    }
        public String toString(){
		    return this.ToString();
	    }
        
    }
}

File: calendar2_TokenRing_Algorithm/HPN JAVA TOKEN RING CODE/HpnCalendarTools/src/hpn/apache/xml/client/HostUrl.java
package hpn.apache.xml.client;

import hpn.settings.DefaultPort;
import hpn.settings.MachinIdentification;

import java.net.MalformedURLException;
import java.net.URI;
import java.net.URISyntaxException;
import java.net.URL;

import org.apache.commons.validator.routines.UrlValidator;

public class HostUrl {
	private String hostUrlAddress;
	private int port;
	private long hostId;
	public HostUrl(String hostUrl) throws IllegalArgumentException
	{
		super();
		this.setHostUrl(hostUrl);
		this.setPort(DefaultPort.portNumber); //Default port for all hostess is 8080.
		this.setHostId();
	}
	public HostUrl(String hostUrl, int port) throws IllegalArgumentException
	{
		super();
		this.setHostUrl(hostUrl);
		this.setPort(port);
		this.setHostId();
	}
	public HostUrl(int port, String ipv4Address) throws IllegalArgumentException
	{
		super();
		this.setHostUrl("http://"+ipv4Address+"/");
		this.setPort(port);
		this.setHostId();
	}
	protected HostUrl(int port) 
	{
		super();
		this.setHostUrl("http://"+MachinIdentification.getIpAddress()+"/");  //For sending requests to the current machine
		this.setPort((port > 1024 && port<65535)? port : MachinIdentification.getPort()); //Default port for all hostess is 8080.
		this.setHostId();
	}
	public String getFullUrl() throws MalformedURLException
	{
        try
        {
        	URL url = new URI(this.hostUrlAddress).normalize().toURL();
        	return url.getProtocol() + "://" + url.getHost() + ":" + this.port + "/";
        }
        catch (URISyntaxException e) 
        {
            throw new MalformedURLException(e.getMessage());
        }
	}
	public String getHostUrl() 
	{
		return this.hostUrlAddress;
	}
	public void setHostUrl(String hostUrlAddress) throws IllegalArgumentException
	{
		UrlValidator urlValidator = new UrlValidator();
		if (urlValidator.isValid(hostUrlAddress)) 
		{
			this.hostUrlAddress = hostUrlAddress;
		} 
		else 
		{
		       throw new IllegalArgumentException("The URL address of the host ["+hostUrlAddress+"] is invalid.");
		}
	}
	public int getPort() 
	{
		return this.port;
	}
	public void setPort(int port) throws IllegalArgumentException 
	{
		if(port>1024 && port<= 65535)
		{
			this.port = port;
		}
		else 
		{
		       throw new IllegalArgumentException("The port number of the host ["+port+"] is invalid. It must be between 1025 and  65535.");
		}
	}
	public long getHostId() 
	{
		return hostId;
	}
	private void setHostId()
	{
		String str = "";
		try {
			URL url = new URI(this.hostUrlAddress).normalize().toURL();
			String[] parts = url.getHost().split("\\.");
			str =  parts[0]+parts[1]+parts[2]+parts[3]+this.getPort();
			this.hostId = Long.parseLong(str);
		} catch (Exception e) {
			this.hostId = -1;
			System.out.println("couldn't calculate host Id. : " + str);
		}
		
	}
	public int compare(HostUrl o2)
	{
		if(this.getHostId()<o2.getHostId())
			return -1;
		else if(this.getHostId()>o2.getHostId())
			return 1;
		else
			return 0;
	}
	@Override
	public String toString(){
		return "URL:["+this.getHostUrl()+"] Port:["+this.getPort()+"]\n";
	}

}
